File: guessingcircle/submissions/accepted/per_n2_maybe_correct.py
#!/usr/bin/env python3

def in_ival(l, x, r):
    if r < l: r += n
    if x < l: x += n
    return x <= r;

def overlap(I, J):
    (l1, r1) = I
    (l2, r2) = J
    if l1 == -1 or l2 == -1: return False
    if r1 < l1: r1 += n
    if r2 < l2: r2 += n
    if r1 < l2:
        l1 += n
        r1 += n
    if r2 < l1:
        l2 += n
        r2 += n
    return r1 >= l2

def fix_ival(l, r):
    l %= n
    r %= n
    while l != r and not in_ival(l, prev[l], r): l = (l+1) % n
    while l != r and not in_ival(l, next[r], r): r = (r-1) % n
    if not in_ival(l, prev[l], r): return (-1, 0)
    return (l, r)

n = int(input())
circle = list(map(int, input().split()))

num_idx = {}
positions = []
for (i, c) in enumerate(circle):
    if c not in num_idx:
        num_idx[c] = len(num_idx)
        positions.append([])
    positions[num_idx[c]].append(i)

m = len(num_idx)
prev = [0]*n
next = [0]*n
for i in range(m):
    x = positions[i][-1]
    for y in positions[i]:
        prev[y] = x
        next[x] = y
        x = y

excludeL = [(-1, 0)]*m
excludeR = [(-1, 0)]*m

for i in range(m):
    if len(positions[i]) == 1:
        x = positions[i][0]
        excludeL[i] = fix_ival(x - (n-1)//2, x - 1)
        excludeR[i] = fix_ival(x + 1, x + (n-1)//2)
    else:
        x = positions[i][-1]
        for y in positions[i]:
            if (y-x) % n > n//2 + 1:
                excludeL[i] = fix_ival(x + n//2 + 1, y - 1)
                excludeR[i] = fix_ival(x + 1, y - n//2 - 1)
                break
            x = y

ans = []
for (num, i) in num_idx.items():
    exL = excludeL[i]
    exR = excludeR[i]
    for j in range(m):
        if not (j == i or overlap(exL, excludeR[j]) or overlap(exR, excludeL[j])):
            break
    else:
        ans.append(num)

if not ans:
    print("none");
else:
    ans.sort()
    print(*ans)

File: guessingcircle/submissions/accepted/per_n2_maybe_correct.cc
#include<bits/stdc++.h>
using namespace std;

typedef unsigned long long word;

int main(void) {
    int n;
    scanf("%d", &n);
    vector<int> circle(n);
    unordered_map<int, int> num_idx;
    vector<vector<int>> positions;
    for (int i = 0; i < n; ++i) {
        scanf("%d", &circle[i]);
        if (num_idx.find(circle[i]) == num_idx.end()) {
            num_idx[circle[i]] = num_idx.size();
            positions.push_back({});
        }
        positions[num_idx[circle[i]]].push_back(i);
    }
    auto in_ival = [n](int l, int x, int r) {
                       while (r < l) r += n;
                       while (x < l) x += n;
                       return x <= r;
                   };
    auto overlap = [n](pair<int, int> I, pair<int, int> J) {
                       if (I.first == -1 || J.first == -1) return false;
                       while (I.second < I.first) I.second += n;
                       while (J.second < J.first) J.second += n;
                       while (I.second < J.first) I.first += n, I.second += n;
                       while (J.second < I.first) J.first += n, J.second += n;
                       return I.second >= J.first;
                   };
    int m = num_idx.size();
    vector<int> prev(n), next(n);
    for (int i = 0; i < m; ++i) {
        int x = positions[i].back();
        for (int y: positions[i]) {
            prev[y] = x;
            next[x] = y;
            x = y;
        }
    }
    vector<pair<int,int>> excludeL(m, make_pair(-1, 0));
    vector<pair<int,int>> excludeR(m, make_pair(-1, 0));
    auto fix_ival = [&](int l, int r) {
                        l = (l + n) % n;
                        r = (r + n) % n;
                        while (l != r && !in_ival(l, prev[l], r)) l = (l+1) % n;
                        while (l != r && !in_ival(l, next[r], r)) r = (r+n-1) % n;
                        if (!in_ival(l, prev[l], r)) return make_pair(-1, 0);
                        return make_pair(l, r);
                    };

    for (int i = 0; i < m; ++i) {
        if (positions[i].size() == 1) {
            int x = positions[i].front();
            excludeL[i] = fix_ival(x - (n-1)/2, x - 1);
            excludeR[i] = fix_ival(x + 1, x + (n-1)/2);
        } else {
            for (int b = 0, a = positions[i].size()-1; b < positions[i].size(); a = b++) {
                int x = positions[i][a], y = positions[i][b];
                if ((y-x+n) % n > n/2+1) {
                    excludeL[i] = fix_ival(x + n/2 + 1, y - 1);
                    excludeR[i] = fix_ival(x + 1, y - n/2 - 1);
                    break;
                }
            }
        }
    }

    vector<int> ans;
    for (auto p: num_idx) {
        auto &[num, i] = p;
        bool ok = true;
        auto exL = excludeL[i], exR = excludeR[i];

        for (int j = 0; ok && j < m; ++j) {
            if (j == i) continue;
            bool diff = overlap(excludeL[i], excludeR[j]) || overlap(excludeR[i], excludeL[j]);
            ok &= diff;
        }
        if (ok) ans.push_back(num);
    }
    if (ans.empty())
        printf("none\n");
    else {
        sort(ans.begin(), ans.end());
        for (int x: ans)
            printf("%d\n", x);
    }
}
